test/units/parsing/yaml/test_dumper.py

```diff
@@ -59,3 +60,40 @@ def test(self):
         data_from_yaml = loader.get_single_data()
 
         self.assertEqual(plaintext, data_from_yaml.data)
+
+    def test_bytes(self):
+        b_text = b'some bytes'
+        f = AnsibleUnsafeBytes(b_text)
+        yaml_out = self._dump_string(f, dumper=self.dumper)
+
+        stream = self._build_stream(yaml_out)
+        loader = self._loader(stream)
+
+        data_from_yaml = loader.get_single_data()
+
+        self.assertEqual(b_text, data_from_yaml)
+
+
+    def test_unicode(self):
+        u_text = u'some unicode'
+        f = AnsibleUnsafeText(u_text)
+        yaml_out = self._dump_string(f, dumper=self.dumper)
+
+        stream = self._build_stream(yaml_out)
+        loader = self._loader(stream)
+
+        data_from_yaml = loader.get_single_data()
+
+        self.assertEqual(u_text, data_from_yaml)
+
+    def test_native_text(self):
+        n_text = 'some native text'
+        f = AnsibleUnsafeText(n_text)
+        yaml_out = self._dump_string(f, dumper=self.dumper)
+
+        stream = self._build_stream(yaml_out)
+        loader = self._loader(stream)
+
+        data_from_yaml = loader.get_single_data()
+
+        self.assertEqual(n_text, data_from_yaml)
```

Comment:
I'm not sure this should work (that may be a "this shouldn't work in the future even if it works now")  I think AnsibleUnsafeText should convert any string it's given into a text string, ie: unicode type on python2 and str type on python3.

This may be the case now but the self.assertEqual is just not picking it up.  For each of your three test cases, try putting non-ascii characters into them.

(Note that the bytes one will needs a special incantation so it works on python3:
``` python
b_text = u'some byte string café'.encode('utf-8')
```

Reply:
Non-ASCII characters in Python 3 pass tests:
```
 0.1s test_bytes: data_from_yaml=b'some byt\xc3\xa9s',
      type(data_from_yaml)=<class 'bytes'>

 0.1s test_native_text: data_from_yaml='some nativé text',
      type(data_from_yaml)=<class 'ansible.parsing.yaml.objects.AnsibleUnicode'>

 0.1s test_unicode: data_from_yaml='some unicodé',
      type(data_from_yaml)=<class 'ansible.parsing.yaml.objects.AnsibleUnicode'>
```

Python 2 fails with non-ASCII characters:
```
# pytest output
self = <units.parsing.yaml.test_dumper.TestAnsibleDumper testMethod=test_bytes>
>       self.assertEqual(b_text, data_from_yaml)
E       AssertionError: 'some byt\xc3\xa9s' != u'some byt\xe9s'

self = <units.parsing.yaml.test_dumper.TestAnsibleDumper testMethod=test_native_text>
>       f = AnsibleUnsafeText(n_text)
E       UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 10: ordinal not in range(128)

# some q output
 0.1s test_bytes: data_from_yaml=u'some byt\xe9s',
      type(data_from_yaml)=<class 'ansible.parsing.yaml.objects.AnsibleUnicode'>

 0.1s test_unicode: data_from_yaml=u'some unicod\xe9',
      type(data_from_yaml)=<class 'ansible.parsing.yaml.objects.AnsibleUnicode'>
```

Reply:
Yep... Reading the Unsafe code, AnsibleUnsafeText and AnsibleUnsafeBytes are just light wrappers around the equivalent python types.  Since the equivalent python types are fragile when mixing types, this is fragile as well.  I think we need to consider it invalid to pass Text type to AnsibleUnsafeBytes and invalid to pass Bytes type to AnsibleUnsafeText.  By extension, it's also invalid to pass Native strings to either AnsibleUnsafeText or AnsibleUnsafeBytes because that will be wrong on one of python2 or python3.